File: app/src/main/java/com/riprog/launcher/data/local/db/entity/HomeItemEntity.java
package com.riprog.launcher.data.local.db.entity;

import androidx.room.Entity;
import androidx.room.PrimaryKey;
import com.riprog.launcher.data.model.HomeItem;

@Entity(tableName = "home_items")
public class HomeItemEntity {
    @PrimaryKey(autoGenerate = true)
    public long id = 0;
    public Long parentId = null;
    public String type;
    public String packageName = null;
    public String className = null;
    public String folderName = null;
    public float row;
    public float col;
    public int spanX;
    public int spanY;
    public int page;
    public int widgetId;
    public float rotation;
    public float scale;
    public float tiltX;
    public float tiltY;

    public HomeItem toDomainModel() {
        HomeItem item = new HomeItem();
        item.setType(HomeItem.Type.valueOf(type));
        item.setPackageName(packageName);
        item.setClassName(className);
        item.setFolderName(folderName);
        item.setRow(row);
        item.setCol(col);
        item.setSpanX(spanX);
        item.setSpanY(spanY);
        item.setPage(page);
        item.setWidgetId(widgetId);
        item.setRotation(rotation);
        item.setScale(scale);
        item.setTiltX(tiltX);
        item.setTiltY(tiltY);
        return item;
    }

    public static HomeItemEntity fromDomainModel(HomeItem item, Long parentId) {
        HomeItemEntity entity = new HomeItemEntity();
        entity.parentId = parentId;
        entity.type = item.getType() != null ? item.getType().name() : HomeItem.Type.APP.name();
        entity.packageName = item.getPackageName();
        entity.className = item.getClassName();
        entity.folderName = item.getFolderName();
        entity.row = item.getRow();
        entity.col = item.getCol();
        entity.spanX = item.getSpanX();
        entity.spanY = item.getSpanY();
        entity.page = item.getPage();
        entity.widgetId = item.getWidgetId();
        entity.rotation = item.getRotation();
        entity.scale = item.getScale();
        entity.tiltX = item.getTiltX();
        entity.tiltY = item.getTiltY();
        return entity;
    }
}

File: app/src/main/java/com/riprog/launcher/data/local/db/entity/HomeItemEntity.kt
package com.riprog.launcher.data.local.db.entity

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.riprog.launcher.data.model.HomeItem

@Entity(tableName = "home_items")
data class HomeItemEntity(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val parentId: Long? = null,
    val type: String,
    val packageName: String? = null,
    val className: String? = null,
    val folderName: String? = null,
    val row: Float,
    val col: Float,
    val spanX: Int,
    val spanY: Int,
    val page: Int,
    val widgetId: Int,
    val rotation: Float,
    val scale: Float,
    val tiltX: Float,
    val tiltY: Float
) {
    fun toDomainModel(): HomeItem {
        val item = HomeItem()
        item.type = HomeItem.Type.valueOf(type)
        item.packageName = packageName
        item.className = className
        item.folderName = folderName
        item.row = row
        item.col = col
        item.spanX = spanX
        item.spanY = spanY
        item.page = page
        item.widgetId = widgetId
        item.rotation = rotation
        item.scale = scale
        item.tiltX = tiltX
        item.tiltY = tiltY
        // folderItems will be populated by the repository
        return item
    }

    companion object {
        fun fromDomainModel(item: HomeItem, parentId: Long? = null): HomeItemEntity {
            return HomeItemEntity(
                parentId = parentId,
                type = item.type?.name ?: HomeItem.Type.APP.name,
                packageName = item.packageName,
                className = item.className,
                folderName = item.folderName,
                row = item.row,
                col = item.col,
                spanX = item.spanX,
                spanY = item.spanY,
                page = item.page,
                widgetId = item.widgetId,
                rotation = item.rotation,
                scale = item.scale,
                tiltX = item.tiltX,
                tiltY = item.tiltY
            )
        }
    }
}
